Index out of bounds when printing large body to HTTP response on Ubuntu in release mode

## Bug Report

Creating on behalf of @jwoertink.

```cr
require "socket"
require "http/server/response"

server = TCPServer.new "0.0.0.0", 8081
spawn do
  if io = server.accept?
    io.sync = false

    r = HTTP::Server::Response.new(io)

    r.print "a" * 45_043
  end
end

client = TCPSocket.new("0.0.0.0", 8081)
client << "message\n"
```

```sh
$ sudo docker run --rm -it -v $PWD:/app -w /app crystallang/crystal:1.7.0 crystal run --release test.cr 
Unhandled exception in spawn: Index out of bounds (IndexError)
  from /usr/share/crystal/src/crystal/spin_lock.cr:25:5 in 'unbuffered_write'
  from /usr/share/crystal/src/http/server/response.cr:263:11 in 'unbuffered_write'
  from test.cr:11:5 in '->'
  from /usr/share/crystal/src/fiber.cr:146:11 in 'run'
  from ??
```

## Reproduction Notes

This one is quite the bear to reproduce. This is the best I got so far, but wanted to get the issue created.

* Only reproduces within an Ubuntu docker image, such as `crystallang/crystal:latest`
  * I also tried `ubuntu:22.04` and manually installed Crystal and same thing
  * I'd also assume it happens on an actual Ubuntu desktop install
  * Doesn't reproduce on mac, or my Arch based desktop
* Only happens in `--release` mode
* Starts happening in Crystal `1.7.0`
  * Versions earlier than that do not reproduce
  * Both `1.6.2` and `1.7.0` are on LLVM `13.0.1`
  * I'll try a bisect to see if i can figure out what commit it started at
* Printing `a` fewer than `45043` times does not produce the error
* Passing a non `IO::Buffered` `IO` type to `HTTP::Server::Response` does not produce the error
  * E.g. `IO::Memory` or `STDOUT`
  * Creating an empty custom `IO` type that includes `IO::Buffered` doesnt change anything either
  * The `IO::Buffered` type must have `.sync = false` and `.flush_on_newline = false` as well to reproduce

## Current Thoughts

Given it only happens with non-synced, when not flushing on new lines, i'm thinking it just has to be something related to https://github.com/crystal-lang/crystal/blob/master/src/io/buffered.cr#L150-L160. Where the size of the response triggering something in one of the one of those like `slice[0, index/count]` calls. E.g. hitting up like https://github.com/crystal-lang/crystal/blob/master/src/slice.cr#L213 which results in the `IndexError`

Well, I spent several hours trying to figure this out, but I can't get an old enough version of Crystal to run locally in order to bisect it. The last version it worked on isn't 1.6.2, it's 1.1.1. So some commit after 1.1.1 is where it broke. The earliest version I was able to run locally is 1.3.0, and that still has the error. 

I also tried to get this to run in a Docker container, but I kept running in to all kinds of random issues like `Error: can't find file 'prelude'` and `undefined reference to `EVP_CIPHER_block_size'` and a few LLVM related issues. If someone is able to create a Dockerfile I can just drop in and run, I can test it, but I think this is beyond my skillset at this point.

--- a/src/lib_c/i386-linux-gnu/c/sys/socket.cr
+++ b/src/lib_c/i386-linux-gnu/c/sys/socket.cr
@@ -54,10 +54,10 @@ lib LibC
   fun getsockname(fd : Int, addr : Sockaddr*, len : SocklenT*) : Int
   fun getsockopt(fd : Int, level : Int, optname : Int, optval : Void*, optlen : SocklenT*) : Int
   fun listen(fd : Int, n : Int) : Int
-  fun recv(fd : Int, buf : Void*, n : Int, flags : Int) : SSizeT
-  fun recvfrom(fd : Int, buf : Void*, n : Int, flags : Int, addr : Sockaddr*, addr_len : SocklenT*) : SSizeT
-  fun send(fd : Int, buf : Void*, n : Int, flags : Int) : SSizeT
-  fun sendto(fd : Int, buf : Void*, n : Int, flags : Int, addr : Sockaddr*, addr_len : SocklenT) : SSizeT
+  fun recv(fd : Int, buf : Void*, n : SizeT, flags : Int) : SSizeT
+  fun recvfrom(fd : Int, buf : Void*, n : SizeT, flags : Int, addr : Sockaddr*, addr_len : SocklenT*) : SSizeT
+  fun send(fd : Int, buf : Void*, n : SizeT, flags : Int) : SSizeT
+  fun sendto(fd : Int, buf : Void*, n : SizeT, flags : Int, addr : Sockaddr*, addr_len : SocklenT) : SSizeT
   fun setsockopt(fd : Int, level : Int, optname : Int, optval : Void*, optlen : SocklenT) : Int
   fun shutdown(fd : Int, how : Int) : Int
   fun socket(domain : Int, type : Int, protocol : Int) : Int

--- a/src/lib_c/x86_64-linux-gnu/c/sys/socket.cr
+++ b/src/lib_c/x86_64-linux-gnu/c/sys/socket.cr
@@ -54,10 +54,10 @@ lib LibC
   fun getsockname(fd : Int, addr : Sockaddr*, len : SocklenT*) : Int
   fun getsockopt(fd : Int, level : Int, optname : Int, optval : Void*, optlen : SocklenT*) : Int
   fun listen(fd : Int, n : Int) : Int
-  fun recv(fd : Int, buf : Void*, n : Int, flags : Int) : SSizeT
-  fun recvfrom(fd : Int, buf : Void*, n : Int, flags : Int, addr : Sockaddr*, addr_len : SocklenT*) : SSizeT
-  fun send(fd : Int, buf : Void*, n : Int, flags : Int) : SSizeT
-  fun sendto(fd : Int, buf : Void*, n : Int, flags : Int, addr : Sockaddr*, addr_len : SocklenT) : SSizeT
+  fun recv(fd : Int, buf : Void*, n : SizeT, flags : Int) : SSizeT
+  fun recvfrom(fd : Int, buf : Void*, n : SizeT, flags : Int, addr : Sockaddr*, addr_len : SocklenT*) : SSizeT
+  fun send(fd : Int, buf : Void*, n : SizeT, flags : Int) : SSizeT
+  fun sendto(fd : Int, buf : Void*, n : SizeT, flags : Int, addr : Sockaddr*, addr_len : SocklenT) : SSizeT
   fun setsockopt(fd : Int, level : Int, optname : Int, optval : Void*, optlen : SocklenT) : Int
   fun shutdown(fd : Int, how : Int) : Int
   fun socket(domain : Int, type : Int, protocol : Int) : Int

